### ISSUE TITLE
[Feat] Allow to animate transitions between routes?

### ISSUE THREAD

**Issue by luchillo17** [angular/angular#9845]
**I'm submitting a ...**  (check one with "x")
- [x] feature request

**Current behavior**
Router replaces the component that is being shown.

**Expected/desired behavior**
Allow for keeping the current component for a configurable ammount of time and instantiating the next component just under so we can animate the transition between those routes.

**What is the motivation / use case for changing the behavior?**
Allowing to transition between routes with animations is just fancy but allows for smoother transitions between pages, it feels bad when your app needs to change a lot of pages or routes and it does it instantly, is unpleasant to the eye.
- **Angular version:** 2.0.0-rc.4
- **Browser:** all
- **Language:** TypeScript

**Comment by akomm** [comment]
Assume you have route `target`. There are different ways of accessing `target`. There might be a variable `source` route from which you come, but there might also be none (initial open in browser via url input or maybe history-api). 

In which cases would you like to have a "transition" to `target` ?

Feels like a responsibility of a component for example, which could handle click to navigate via router-api, after completing an transition - e. G. using reactivex.

**Comment by Luchillo** [comment]
Soo it should be handled like a controller that intercepts the route request and animates (say a fade off) the current page then calls the router api to trigger the new route and animates the new route into the view?

If that's what you say (at least that's what i understood) there's 2 things that doesn't work any more, the router's guards (the ones you attach to canActivate and canDeactivate) that are attached in the routes, not in the component itself and the animation of both components have to be serial, the 1º animation has to finish before the 2º one can start since those 2 routes can't exist at the same time.

**Comment by zoechi** [comment]
You could use a custom router-outlet that adds °2 and then removes 1° only after the animation is completed until this will be supported out of the box.

**Comment by Luchillo** [comment]
I'm not going to make this from scratch, my current project supports lazy loading with the router v3 and i'm not going to risk losing that feature by making a router by my own.

I think this should be supported, the question is if it will ever be.

**Comment by zoechi** [comment]
You can just copy or extend the current router-outlet and manipulate the createComponent/destroy logic a bit. There isn't much going on in router-outlet. If you rather wait until it is provided out-of-the box it's fine as well of course.

**Comment by Luchillo** [comment]
I have more things to think about in my current apps, once i have time i'll try to make just that (extending the router), meanwhile i'll wait for the Router's native support.

**Comment by matsko** [comment]
So I have two pieces of news for you:

1. When RC5 rolls out, it is possible to animate routes via the `host: { }` properties on a component.

```ts
@Component({
   host: {
    // it doesn't matter what the expression value is since we only care about `*` and `*void`
     '[@routeAnimation]': "true"
   },
   animations: [
      trigger('routeAnimation', [
         transition('void => *', [
             // when the component enters
         ]),
         transition('* => void', [
             // when the component leaves
         ])
      ])
   ]
})
class MyRoutedComponent {

}
```

For each component that is routed you'll need to attach the host animation property binding and the trigger + transitions.

2. Now once we get `query()` support into animations, sometime in a few weeks or so, we can do this:

```ts
// app.html
<router-outlet @routerAnimations="routerExp"></router-outlet>

// app.ts
@Component({
   animations: [
     trigger('routerAnimations', [
        transition('/home => /about', [
           query('$previous', [
              // animate the old page in 
           ]),
           query('$next', [
              // animate the new page in 
           ])
        ]),
        transition('/about => /forums', [ ... ]),
        transition('* => /home', [ ... ]),
     ])
   ]
})
class App {
}
```

**Comment by Luchillo** [comment]
If that's so maybe transitions could be defined with the `void` state? say defining transitions from void and to void like:
```js
@Component({
   animations: [
     trigger('routerAnimations', [
        transition('* => void', [ // For previous page
        ]),
        transition('void => *', [ ... ]) // For next page
     ])
   ]
})
```

**Comment by matsko** [comment]
@Luchillo that should work in RC5 (don't forget about referencing `routerAnimations` on the component `host:` data).

**Comment by Luchillo** [comment]
Sorry, i missed the point, i have to reference the `routerAnimations` in the components that works as route, and as host?

I'm a bit lost there, do you mean this part?
```js
// app.html
<router-outlet [@routerAnimations]="myOutlet.state" #myOutlet></router-outlet>
```

**Comment by Luchillo** [comment]
I should've explained myself better, with the `Full-Fledge` api could we define animation in the outlet for all pages using `void => *` and `* => void` for entering and leaving animations?

If so is it mandatory to also define the animation in each page component?

**Comment by matsko** [comment]
RC5 will hopefully be out this week, if not then next week.

For now (with RC5) you will need to do this for **every component** that is routed to and you want to add animations to:

```ts
@Component({
   host: {
     '[@routeAnimation]': 'true'
   },
   animations: [
      trigger('routeAnimation', [
         transition('* => void', animate(...)),
         transition('void => *', animate(...))
      ])
   ]
})
class Cmp { }
```

Once we get 'query()` and `$variables` into animations then you can use `<router-outlet>` with the URL API that I wrote above.

**Comment by MadUser** [comment]
Hi @matsko, there is a bug in this approach when using nested router.
I have animation on my father page which has a router itself.
when the child router change state (no animation declared on child pages)
I get the animation running on the father page all over again.

**Comment by matsko** [comment]
hey @MadUser. Any chance you can put together a demo of this so that I can investigate? Here is a starter plnkr to get your started: http://plnkr.co/edit/414ar3acXpHymiPrSeun?p=preview

**Comment by MadUser** [comment]
I created an example based on @andrui plunker
from this issue: https://github.com/angular/angular/issues/9350
and learned two new things:
1. this animation demo doesn't work on IE11.
2. the animation glitch that I am talking about happens only at the first time.
here it is:
http://plnkr.co/edit/c41pebxqPQpKrJg6cHHq?p=preview

**Comment by flauc** [comment]
Hi @matsko, i've been playing around with `'[@routeAnimation]': 'true'` and it works great. Now i'm trying to only animate page transitions and not the initial load. Is there maybe a solution in place for this already?

**Comment by Luchillo** [comment]
@flauc If you think about it, **i think** the starting animation happens because the way the transitions are declared states that when first created and when destroyed should have animation:
```js
      trigger('routeAnimation', [
         transition('* => void', animate(...)), // Here we define animation when leaving and destroying
         /* Next we define from void state to any, which i think determinates the animation should run
          * when the element is added or created, which is the case of page transition since we are 
          * replacing the page with another page.
         transition('void => *', animate(...))
      ])
```

So as both initial load and page transitions perform the same operation of creating or replacing the page component, the animation will happen always, so if you really need to avoid animation in first load **i think** you will have to wait until query is supported by the router as stated in comment up in this thread.
https://github.com/angular/angular/issues/9845#issuecomment-232476173

**Comment by matsko** [comment]
@flauc yes we have thought about this. Me and Misko do have an idea in mind, but the API needs a little bit more work. We are trying to have something in place for the next release after RC6.

**Comment by flauc** [comment]
That's great to hear thank you  :+1:

**Comment by born2net** [comment]
wondering if this made it into rc.7?

**Comment by andrui** [comment]
@born2net 
https://github.com/angular/angular/milestones  there is no rc.7 scheduled. Next is 2.0 release

**Comment by matsko** [comment]
@born2net @andrui no this didn't make it into RC7: https://github.com/angular/angular/blob/master/CHANGELOG.md#200-rc7-2016-09-13

For this feature to work `query()` needs to be developed (I am currently working on that now). Hopefully we can have something functional in the coming weeks.

**Comment by eXxi** [comment]
Unfortunately, the leave (* => void) animations are not working in child components (only the parent animation is applied). Are there any quick fixes / workarounds?

**Comment by gustavolira** [comment]
Any news about this feature?

**Comment by regenrek** [comment]
Seems there is some progress on this: https://github.com/matsko/angular-connect-demos

**Comment by andrui** [comment]
Hi @MadUser 

I have the similar problem. Did you find any suitable solution for child nested routes?

**Comment by andrui** [comment]
I just was able to implement artificial solution with "Dummy Component": http://plnkr.co/edit/mSz2fqjTzR69wp3FKuzP?p=preview

**Comment by andrui** [comment]
But in general my problem is bit another. I need to animate re-usable "card" component during route chanages. E.g. I have a tree of routable components:

**|-App**
....**|- Page 1**
........**|- Page 1.1**
............**|- Page 1.1.1**
............**|- Page 1.1.2**
........**|- Page 1.2**
....**|- Page 2**
........**|- Page 2.1**

With routing accordingly:
/page1
/page1/page1_1
/page1/page1_1/page1_1_1
/page1/page1_1/page1_1_2
/page1/page1_2
/page2
/page1/page2_1

Template for each of these components consists of: 
```(html)
<my-card-cmp>some page content here</my-card-cmp>
<router-outlet><router-outlet>
```

I need to have same animation for "my-card-cmp" during component creation/destruction when route is  changed (e.g. transform scale from 0 to 1 and vice versa). So I expect to have behaviour like:
1) I go to /page1/page1_1/page1_1_1 => all "my-card-cmp" from each page component (Page 1, Page 1.1, Page 1.1.1) are scaled from 0 to 1;
2) then on page "page1" I press link to page1_2 aka got to /page1/page1_2 => all "my-card-cmp"  from components which are going to be destroyed by route change (Page 1.1, Page 1.1.1) are scaled from 1 to 0 and then "my-card-cmp" from Page 1.2 is scaled from 0 to 1

@matsko 
- is it possible to implement such animation behaviour somehow with latest angular version?
- will it be possible to implement this with future `query()` functionality?
- is there any estimates for `query()` push request into main angular repo?

Thank you in advance!

**Comment by apur27** [comment]
Hello everyone!

First of all the suggestions above were helpful but i am afraid facing some issues implementing smooth page transition in Angular 2

I will highly appreciate if anyone here can share their thoughts around these 

A.  We currently have a background component which we are not able to configure based on page transition as it is outside router-outlet, whenever a route change happens we change the file name of the background image.
<header role="banner" class="banner">
</header>
<main role="main" lb-retain-scroll>
  <router-outlet></router-outlet>
</main>
<jk-background></jk-background>

B. During the route change we see the content the current route leaving below the screen while the new content arriving from the top. What should ideally happen is the current screen to slide left and new screen to slide left from the right hand side.
<------------------------------------sliding left
<----- current route -------> <----- new route -------> 
                         |<----- screen display ------->|

C.  is there a way out-of-the-box to order a route parent-child-grandChild where by if you go from 
Parent to Child
<------------------------------------sliding left
<----- Parent  child -------> <----- child route -------> 
                         |<----- screen display ------->|
and if you go from Child to Parent
                                                    sliding right------------------------------------>
<----- Parent  child -------> <----- child route -------> 
                         |<----- screen display ------->|

**Comment by francosioquim** [comment]
You could use a custom router-outlet that adds °2 and then removes 1° only  
after the animation is completed until this will be supported out of the  
box.

---

**Comment by born2net** [comment]
the feature is now supported in 2.1:
https://github.com/matsko/angular-connect-demos

**Comment by andrui** [comment]
@born2net what exact feature do you mean? In https://github.com/matsko/angular-connect-demos there is a function which just returns animation trigger which is used for the host binding of **routable** component. I am interested in functionality of child components animation. Will `query()` provide this functionality? E.g. in context of demo_1 from "angular-connect-demos" is it possible to collapse content before route is changed?

**Comment by lazarljubenovic** [comment]
The readme from Matias's demo repo just states that "the animations that use `query()` (demo 4) are using a custom API that is still being developed". Also [here's](https://youtu.be/4VDmIVrML34?t=910) Matias talking about `query()`.

In his demo repo, there's [`AnimationQueryAst`](https://github.com/matsko/angular-connect-demos/blob/7d149dabc31ad748b802b65d5330d09eefadb170/%40angular/compiler/src/animation/animation_ast.d.ts#L87) but no such thing is yet available in [the same file in 2.1-rc0](https://github.com/angular/angular/blob/2.1.0-rc.0/modules/%40angular/compiler/src/animation/animation_ast.ts). The changelog for 2.1-rc0 also only mentions animations from [this commit](https://github.com/angular/angular/pull/11991/commits/695e87fbd3844086ccca4ee52dc5eedd14237752) which introduces aliases for `:enter` and `:leave` which can be used instead of `void => *` and `* => void`. I don't see anything about `query()`.

I don't remember hearing anything specific about it coming in 2.1 from the conference video either. Not sure where you got that information from, @born2net.

**Comment by born2net** [comment]
yes I was referring to the routing demo, and so yes it is under development as part of animation...
```
 templateUrl: './demo_1/demo.html',
    animations: [routerAnimations('routeAnimations')]
```
so hoping it's coming soon,

regards

**Comment by gustavolira** [comment]
any news?

**Comment by matsko** [comment]
`query` is almost there. We will have an earlier API that will allow for finding the route view containers that are being inserted and removed.

We're trying to get it to work like so:

```ts
trigger('routerAnimations', [
  transition('/home => *', [
    query(':enter', style({ transform: 'translateX(-100%)' }),
    query(':leave', [
       animate(1000, style({ transform: 'translateX(100%)' }))
    ])
    query(':enter', animate(1000))
  ])
])
```

We're also looking into getting an API together that allows for child components to animate as well within the context of a view change.

**Comment by apur27** [comment]
One quick question

What if there is a background component which controls background images which is outside routerOutlet

**Comment by andrui** [comment]
@matsko 

Thank you for the answer. Can you pls just clarify one thing: can  `query` be used for child reusable components animation as I described here https://github.com/angular/angular/issues/9845#issuecomment-251940189 ?

**Comment by zoechi** [comment]
@apur27 I guess it's about [CONTRIBUTING - Got a Question or Problem?](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question)

**Comment by joshking** [comment]
One thing I noticed is that components may have to be absolutely positioned in the DOM to animate out. Animating in doesn't seem to be an issue. 
`  host: {
   '[@routeAnimation]': 'true',
   '[style.display]': "'block'",
   '[style.position]': "'absolute'",
   '[style.left]' : "'0'",
   '[style.right]' : "'0'"
  },`

**Comment by marsdk** [comment]
@matsko  Do you know when the query() will be added to the Angular2 and ready for use - are you missing 3 days or 300 days ? :)

**Comment by marcoskichel** [comment]
Any news about this one?

**Comment by matsko** [comment]
I have a prototype of this working. Currently I'm writing tests and preparing it for review next week.

**Comment by Numel2020** [comment]
@matsko Is it possible to to detect (via callback) when a transition transits into the "enter" or "leave" state?

**Comment by matsko** [comment]
Yes and no. With the query feature you can do the following:

```html
<div [@animation]="state">
   <div #sub [@subAnimation]="subState"
           (@subAnimation.start)="onStart()"
           (@subAnimation.done="onDone()">...</div>
</div>
```

```ts
trigger('animation', [
   transition(':enter', [
      animate(1000, ...)
      query('sub', animateChild())
      animate(1000, ...)
   ])
])
```

The code above will run the `subAnimation` animation, but only when the parent decides. This only works if the elements being animated are within the same component template.

Now there are plans to expand this to work cross component (from parent component to child), so if you think about routing then you could have the sub animation be an animation on the component host binding (right on the component that is being entered into the page) which can act as an enter animation. This way you can find out when the page is entered or removed.

I think in the future having a lifecycle hook for `ngOnAnimate()` would be ideal for this usecase.

**Comment by AKuzmanoski** [comment]
Any news about releasing query feature?

**Comment by ranbuch** [comment]
Until the query feature will be ready I'm animating routes transitions like that:
```
app-root * {
    animation-duration: 1s;
    opacity: 0;
    animation-name: fadeIn;
    animation-fill-mode: forwards;
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
```
Add this code to your ```style.css``` file

**Comment by benjamincharity** [comment]
Good call @ranbuch. In my case I only want the animation on the container for the new view so I've amended it a bit.

```
app-root {

  // animate the first sibling to the router-outlet
  router-outlet {
    & + * {
      animation-duration: 400ms;
      opacity: 0;
      animation-name: viewFadeIn;
      animation-fill-mode: forwards;
    }

  }
```

**Comment by TheAifam5** [comment]
I really like this idea.
Actually don't know how to add animation for every route :(

**Comment by davidalee** [comment]
Any update on release date for the query feature?

**Comment by Delagen** [comment]
@matsko wrote 'Now once we get query() support into animations, sometime in a **few weeks** or so, we can do ...' on **Jul**. Almost half year passed

**Comment by davidalee** [comment]
You wouldn't be required to animate *every* route, you'd have the option of specifying which routes get the transition according to the example noted in https://github.com/angular/angular/issues/9845#issuecomment-232476173.

**Comment by sollabeck** [comment]
I'm guessing that query() support hasn't been implemented yet.

**Comment by davidalee** [comment]
@matsko Can you at least let us know whether this issue is still on the roadmap (albeit without a release date as of yet) or whether it has been scrapped and will not be implemented?

**Comment by bergben** [comment]
Until `query()` is available: I have created a component to do exactly this a few months ago: https://github.com/bergben/ng2-page-transition

**Comment by davidalee** [comment]
Cool, will check that out @bergben. Any idea on how I might implement a different transition depending on the start/end route?

For example, if I have a horizontal menu with 2 buttons and want to implement a left -> right sliding transition when going from page 1 -> page 2 and vice versa.

**Comment by bergben** [comment]
You could do that using https://github.com/bergben/ng2-page-transition#custom-transition (to create your custom transitions) and https://github.com/bergben/ng2-page-transition#onlyonroutes (to only trigger it for the specific route, e.g. [onlyOnRoutes]="['routeWhen/rightButton/isClicked']").  
Feel free to ask anything here: https://gitter.im/bergben/bergben (Let's not spam this thread).

**Comment by dolanmiu** [comment]
Will this ever be supported?

**Comment by bergben** [comment]
Oh I see I am too late @dolanmiu :D

For others, check out Matias ng-conf talk: https://www.youtube.com/watch?v=Oh9wj-1p2BM&t=12m20s
Angular 4.1 will finally support it as it seems ;)

**Comment by luchillo17** [comment]
Well i almost tough i'll have to give up in this feature, request was made 6 of July 2016, almost 9 months ago, after watching the routing animations part of the video i think i can close this issue.

**Comment by dolanmiu** [comment]
It's been one hell of a journey! :D

**Comment by MoLow** [comment]
Well I see on npm that @angular/animations@4.1.0 has been released, but without query  :(

**Comment by Ploppy3** [comment]
It will be release in 4.2.0

**Comment by matsko** [comment]
FYI everyone, the `4.2.0-rc1` has a working version of query that supports routable animations :)

**Comment by AKuzmanoski** [comment]
Great work, I'm really excited to try this feature

However it seams that API has changed from the one you have presented [here](https://www.youtube.com/watch?v=Oh9wj-1p2BM&t=1s)
For example animateChild takes only params, and as i find out in the code, there is other useAnimation function.

Can you please provide us with some API docs

**Comment by Ploppy3** [comment]
@AKuzmanoski even though it's not some docs it's a working example: https://github.com/matsko/ng4-animations-preview (almost up to date)

**Comment by AKuzmanoski** [comment]
Thank you @Ploppy3

**Comment by MrCroft** [comment]
@AKuzmanoski 
Have you managed to understand anything? Or even make something work?

I've posted everywhere, no answer... searched and tried all night. No documentation... nothing.

How does animateChild() work now? When/what do you use it for? How do you tell it to use a certain animation if it doesn't take it as a parameter anymore?
Or do we use `useAnimation()` for that? And how do we use it?
Like this:
```
export const fadeAnimation = trigger('fade', [
    transition('out => in', useAnimation(fadeIn, { params: {customTime: 5000} })),
    transition('in => out', useAnimation(fadeOut, { params: {customTime: 5000} }))
]);
```

Or like this:
```
export const fadeAnimation = trigger('fade', [
    transition('out => in', useAnimation(fadeIn), { params: {customTime: 5000} }),
    transition('in => out', useAnimation(fadeOut), { params: {customTime: 5000}})
]);
```
Because, with any of the above, the cli builds the app just fine, no errors on build.
But, either of the 2 ways above, I get an error in the browser console: `Error: The animation trigger "fade" has failed to build due to the following errors: - The provided timing value "$customTime" is invalid.` for the following piece of code:

```
const fadeIn = animation([
    style({
        opacity: 0
    }),
    animate('$customTime', style({ opacity: 1 }))
], { params: { customTime: 1000 } });

const fadeOut = animation([
    animate('$customTime', style({ opacity: 0.2 }))
], { params: { customTime: 1000 } });
```

The only case I don't get `provided timing value "$customTime" is invalid` is this case:
```
export const fadeAnimation = trigger('fade', [
    transition('out => in', animateChild(fadeIn), { params: {customTime: 5000} }),
    transition('in => out', animateChild(fadeOut), { params: {customTime: 5000}})
]);
```
But, then again, the above code doesn't do anything... it jumps instantly from start to done.
Also, in the demo we simply have `{ customTime: '2s'}` instead of `{ params: { customTime: '2s'} }`... But then I get `customTime does not exist in type AnimationOptions`, because it changed.

I've been up all night after seeing the ng-conf video. It's almost 9am here, and I still couldn't make sense of any of this... Let alone actually make it work :( I'm at wits end.

P.S. I'm on 4.2.0-rc.2 and I'm simply trying this on a div, like in the ng-conf demo (not for routing or anything).
```
<div style="width: 300px; height: 180px; background: green;"
    [@fade]="{ value: visible ? 'in' : 'out', customTime: 1000 }"
    (@fade.start)="animStart($event)"
    (@fade.done)="animEnd($event)"></div>
```
And guess what, I first get **animation end** logged and after that **animation start**. But that's with the `animateChild(fadeIn)`, which jumps instantly from one state to the other, doesn't take any timing into account - even if I set it directly to something, instead of `$customTime`. And I don't even know why doesn't throw the same error as with `useAnimation()` :|

**Comment by matsko** [comment]
@MrCroft sorry for the difficulties you've been facing.

All the features are in master, but there were some syntax changes with the new features just before RC1 was released. Once 4.2 is out (either this week or next), angular.io will be updated with API docs about all the new features and there will be a fully detailed blog post. It's just not out right now because the team might just make last minute changes before the API is stable (once 4.2 is out).

First thing first, the `$var` variables have been changed to be `{{ var }}` and they are defined in `{ params: { var: 'value }}`.

`animateChild` allows for a sub animation trigger (an inner element with a `@trigger` on it) to be animated at the exact right time by the parent. (here is an example of that: https://github.com/angular/angular/issues/15798#issuecomment-305934289). Back during the ngConf presentation it was thought that `animateChild` would be both used for this feature as well as invoking an animation that was created with `animation()`, but this changed and now there is a method called `useAnimation` (which is designed to be used with an animation created with `animation([...])`). The syntax you have above should work:

```
transition('out => in', useAnimation(fadeIn, { params: {customTime: 5000} })),
```

`queryAll` is gone (just use `query` instead).

Once again, as soon as 4.2 is then there will be lots of nice learning material about how all this works.

**Comment by MrCroft** [comment]
@matsko 
Thank you for the detailed answer. It clears things up. So, it is expected that once 4.2 will be released, this will work: `transition('out => in', useAnimation(fadeIn, { params: {customTime: 5000} })),`
That's great news!

**Comment by AKuzmanoski** [comment]
@MrCroft 
`transition('out => in', useAnimation(fadeIn, { params: {customTime: 5000} }))` allready works with rc.1 and rc.2

Here is my working version of FadeIn animation

**Animation Code**
`animation` params with their default values are defined in the second argument after `AnimationMetadata` array
```
export const fadeIn = animation([
  style({
    opacity: 0
  }),
  animate('{{duration}} {{delay}} {{easing}}', style({
    opacity: 1
  }))
], {params: {duration: "158ms", delay: "0ms", easing: "ease-out"}});
```
**Triggerer code**
Here transition takes (name, animation, and animationOptions)
But also this animation options are sent to the animation itself by calling 
`useAnimation(animation, animationOptions)`
The `{{duration}}` param is taken from the transition options
```
trigger("fadeInOut", [
      transition(':enter',
        useAnimation(fadeIn, {params: {duration: "{{duration}}"}}),
      {params: {duration: "150ms"}}),
      transition(':leave', [
        useAnimation(fadeOut)
      ])
    ])
```

**Template Code**
Here we call the animation by its trigerrer
We can send value and also params
Value defines the state
```
<button [@fadeInOut]="{value, {params: {duration: '158ms'}}" md-fab>
```

I hope this was helpful

**Comment by MrCroft** [comment]
@AKuzmanoski It didn't even occur to me to try and use "real" data binding `{{duration}}`. I only tried `'$duration'` as per the demo.
Now, binding a duration value inside the animation code works `animate('{{duration}}', ...`

But when trying to input the duration from my component (`<button [@fadeInOut]="{value, params: {duration: '158ms'}}" md-fab>`), then I get "_The provided timing value "{{duration}}" is invalid._" thrown for  `useAnimation(fadeIn, {params: {duration: "{{duration}}"}})`

**Comment by matsko** [comment]
@MrCroft can you make me a plunkr of this please and I can fix it tomorrow.

**Comment by MrCroft** [comment]
@matsko  Done.
http://plnkr.co/edit/T7mQaDpqWRxf8FCaLKmm?p=preview

**Comment by AKuzmanoski** [comment]
@MrCroft

Just change input parameter name in your transitions from customTime to duration

Your version
```
export const fadeAnimation = trigger('fade', [
    
    transition('out => in',
        useAnimation(fadeIn, { params: {customTime: '{{duration}}'} }),
        {params: {customTime: '5000ms'}}
    ),
    transition('in => out',
        useAnimation(fadeOut, { params: {customTime: '{{duration}}'} }),
        {params: {customTime: '5000ms'}}
    )
]);
```

Working version
Your version
```
export const fadeAnimation = trigger('fade', [
    
    transition('out => in',
        useAnimation(fadeIn, { params: {customTime: '{{duration}}'} }),
        {params: {duration: '5000ms'}}
    ),
    transition('in => out',
        useAnimation(fadeOut, { params: {customTime: '{{duration}}'} }),
        {params: {duration: '5000ms'}}
    )
]);
```

**Comment by AKuzmanoski** [comment]
@MrCroft

I just tried this code in Plunker
It seams that '{{duration}}' value in useAnimation is not parsed

@matsko 
I think that this worked in rc-1

**Comment by AKuzmanoski** [comment]
@matsko
I have problem with animateChild() when routing. Leaving page nested animations dont run, and animated elements just stay in their position while entering page loads.

**Routing animations**
```
trigger('routerAnimations', [
      transition('feed => about',
        group([
          query(':leave',
            animateChild()),
          query(':enter',
            animateChild())
        ])
      )
    ])
```
**Leaving page animation**
The animation is binded to the host element with `@HostBinding("@pageAnimation") animation: boolean = true;`
```
trigger('pageAnimation', [
      transition(':enter', [
        query('@*', animateChild())
      ]),
      transition(':leave', [
        query('@*', animateChild())
      ])
    ])
```

**Inner element animation**
```
trigger("fab", [
      state("void", style({
        transform: "scale(0)"
      })),
      transition(':enter',
        useAnimation(scaleIn)),
      transition(':leave',
        useAnimation(scaleOut))
    ])
```
# Workaround

If I combine the animations together in the page animation like this:
```
trigger('pageAnimation', [
      transition(':enter', [
        query('[md-fab]', useAnimation(scaleIn))
      ]),
      transition(':leave', [
        query('[md-fab]', useAnimation(scaleOut))
      ])
    ])
```
it works just fine.
Maybe the fab button is not aware that page is leaving and `":leave"` transition on the button is not fired.

Should i report this as a new issue?

**Comment by AKuzmanoski** [comment]
@MrCroft

I just accidently find out that you dont have to send params to `useAnimation()`

So code like this finaly works
```
const fadeIn = animation([
    style({
        opacity: 0.2
    }),
    animate('{{ duration }}', style({ opacity: 1 }))
], { params: { duration: '1000ms' } });

const fadeOut = animation([
    animate('{{ duration }}', style({ opacity: 0.2 }))
], { params: { duration: '1000ms' } });

export const fadeAnimation = trigger('fade', [
    
    transition('out => in',[
        useAnimation(fadeIn)],
        {params: {duration: '5000ms'}}
    ),
    transition('in => out',[
        useAnimation(fadeOut)],
        {params: {duration: '5000ms'}}
    )
]);
```
But this aproach is not intuitive i think

**Comment by MrCroft** [comment]
@AKuzmanoski  Hooray :) Thank you!
So, the variable you send from your component **MUST** match the variable you use in your animation.

**This works:** ('5000ms' sent from the component overrides the '1000ms' defined in the animation)
**component:** _[@fade]="{ value, { params: { **duration: '5000ms'** } }"_
**animation:** _transition('out => in', [ useAnimation(fadeIn) ], {params: {**duration: '1000ms'**}} )_

**This doesn't work** (and probably it was never intend to work?):
**component:** _[@fade]="{ value, { params: { **customTime: '1000ms'** } }"_
**animation:** _transition('out => in', [ useAnimation(fadeIn) ], {params: {**duration: '{{ customTime }}'**}} )_

**Comment by MrCroft** [comment]
Sorry to keep posting on this closed issue.
I don't know if this is a bug or not (in which case, there must be a way): It doesn't keep the state after the animation finishes.

For example, for a 'zoom' animation, I used to have the following trigger code:
```
export const magnifyAnimation = trigger('magnify', [
    state('default', style({
        'transform': 'scale(1)'
    })),
    state('magnified', style({
        'transform': 'scale(1.3)'
    })),
    transition('default <=> magnified', [
        animate('430ms ease-out')
    ])
]);
```
Now, with input params, I only have something like this inside my trigger:
```
export const magnifyAnimation = trigger('magnify', [
    transition('default => magnified', [ useAnimation(zoomEffect) ], { params: { scaleFactor: '1.3' } }),
    transition('magnified => default', [ useAnimation(zoomEffect) ], { params: { scaleFactor: '1' } })
]);
```
But as soon as the animation finishes, the element jumps back to it's natural size.
And if I want to add back the `state()` code, it doesn't take input params.
I've also tried this:
```
    state('magnified', style({
        'transform': '*'
    })),
```
but it doesn't do what I would hope (keep the _whatever_ value the transformation had at the end of the animation).

**Comment by matsko** [comment]
@MrCroft https://github.com/angular/angular/pull/17315.

`state()` values purposely (for now) do not accept input params. This is hopefully something to have changed in 4.3.

**Comment by MrCroft** [comment]
@matsko Is there no way to keep the end state, while scale() takes an input param?

**Comment by davidalee** [comment]
You should open a new issue for better tracking/discoverability in case
other users come across the same error.

**Comment by MrCroft** [comment]
Done. #17362

**Comment by Ploppy3** [comment]
Quick question, are they any events to the new router animations like there are in the old one?

**Comment by DBjelovuk** [comment]
So what's the simplest way at this point to get multiple routed components to only animate when transitioned between one another, and not on page load?

**Comment by Ploppy3** [comment]
@DBjelovuk it may look like a lot of code but it's a complete implementation of routing with animations.

app.modules.ts:
You must define every data[animation] value.
```
const routes: Routes = [
  { path: 'page1', component: HomeComponent, data: { animation: 'page-1' } },
  { path: 'page2', component: HomeComponent, data: { animation: 'page-2' } },
];
```
app.component.html:
```
<div [@routerAnimations]="prepareRouteTransition(outlet)">
    <router-outlet #outlet="outlet"></router-outlet>
</div>
```

app.component.ts:
Here you define the 'init' state (when your page load)
```
public prepareRouteTransition(outlet: RouterOutlet) {
    return outlet.activatedRouteData['animation'] || 'init';
}
```

define your animations:
```
export const routerAnimations: AnimationMetadata = trigger('routerAnimations', [
    transition('init=> *', [
        //denies any animation from 'init' state
    ]),
    transition('* => page2', [
        .... // example animation
    ]),
]);
```

In this case an animation would trigger going from any page to page2, but would not trigger when refreshing the page.

**Comment by DBjelovuk** [comment]
Hi Ploppy, thanks for the snappy response! I feel like I'm getting really close to what I want, except I need some way to have my query animations fire in parallel. See: https://stackoverflow.com/questions/45598685/run-multiple-queried-animations-in-parallel

**Comment by Ploppy3** [comment]
@DBjelovuk put your query methods into a group() to execute them in parallel.
Did not check your SO yet.

**Comment by DBjelovuk** [comment]
Ah, so simple. Cheers!

**Comment by sathish94us** [comment]
Will this routing animation work for components which are routed through custom route-reuse-strategy ?? I observed that animation was working only for the first time when normal routing was happening.
